## 1. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 2. Importing data

In [3]:
# Turn project folder path into a string
path = r'C:\Users\Admin\Desktop\Data Analysis\Python\04-2024 Instacart Basket Analysis'

In [4]:
path

'C:\\Users\\Admin\\Desktop\\Data Analysis\\Python\\04-2024 Instacart Basket Analysis'

In [5]:
# Import orders.csv using list
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [7]:
## Droping the unnamed columns which converted from the index from the .csv file
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [8]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
# Import product.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## 3. Mixed-Type Data

In [11]:
# Create a dataframe
df_test = pd.DataFrame()

In [34]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [35]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [36]:
# Check for mixed types in columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
# Changing mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
df_test['mix'].dtypes

dtype('O')

## 4. Missing Data

In [17]:
# Check for missing value in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
# Creating a subset with only null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
df_prods.shape

(49693, 5)

In [21]:
# Creating a subeset without null values from df_prods
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

## 5. Duplicate

In [23]:
## Finding duplicate rows in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
## Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

## 6. Exporting Dataframe

In [49]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## 7. Task

### 7.2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [29]:
# Running describe() on df_ords
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Ans: 
'days_since_prior_order' have less count than the other columns, which could indicate missing data. This could be due the a few following reasons:

    a. Missing data in the 'day_since_prior_order' columns.
    b. Due to new customer not having any 'days_since_prior_order' value.

### 7.3. Check for mixed-type data in your df_ords dataframe.

In [40]:
# Check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [41]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   order_day_of_week       int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


#### Ans: No mixed-type data found.

### 7.4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### Ans: No mixed-type data found.

### 7.5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [42]:
# Check for missing value in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### Ans: 'days_since_prior_order' columns contains 206209 missing values.

### 7.6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [43]:
# Create subset of dataframe that countains the missing value in 'days_since_prior_order'
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [44]:
df_ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [45]:
df_ords_nan.describe()

order_id        user_id  order_number  order_day_of_week  \
count  2.062090e+05  206209.000000      206209.0      206209.000000   
mean   1.708462e+06  103105.000000           1.0           2.754118   
std    9.881299e+05   59527.555167           0.0           2.076205   
min    2.000000e+01       1.000000           1.0           0.000000   
25%    8.507300e+05   51553.000000           1.0           1.000000   
50%    1.706246e+06  103105.000000           1.0           3.000000   
75%    2.564292e+06  154657.000000           1.0           5.000000   
max    3.421081e+06  206209.000000           1.0           6.000000   

       order_hour_of_day  days_since_prior_order  
count      206209.000000                     0.0  
mean           13.626597                     NaN  
std             4.223769                     NaN  
min             0.000000                     NaN  
25%            11.000000                     NaN  
50%            14.000000                     NaN  
75%            17.000000                     NaN  
max            23.000000                     NaN

#### Ans: From the describe function, i can verify that the order number for each of the row which contain NaN for 'days_since_prior_order' is 1, which indicate that the row is the first order transcation of the customer in the store. And for that reason, i believe it is okay to leave all the null value there as any number used to impute into those rows for the 'days_since_prior_order' columns will affect future analysis.

### 7.7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [47]:
# Check for duplicate in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [48]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### Ans: No duplicate was found. 

### 7.8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

#### Ans: No duplicate was found. If duplicate was found, i would create a seperate dataframe called "df_ords_clean" by running the command "df_ords_clean = df_ords.drop_duplicate()" to drop all duplicates from the df_ords dataframe.

### 7.9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [50]:
# Exporting cleaned df_prods as 'products_checked.csv'
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [51]:
# Exporting cleaned df_prods as 'orders_checked.csv'
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))